File: Array3.cpp
//Given an array of N integers left rotate the array by one place : brute force approach : 
// #include<bits/stdc++.h>
// using namespace std;

// void solve(int arr[], int n){
//       int temp[n];
//       for(int i = 1; i < n; i++){
//           temp[ i - 1] = arr[i];
//       }
//       temp[n - 1] = arr[0];
//       for(int i = 0; i < n; i++){
//           cout << temp[i] << " "; 
//       }
//       cout << endl; 
// }
// int main(){
//       int n = 5;
//       int arr[] = {1, 2, 3, 4, 5};
//       solve(arr, n);
// }



// optimal approach: 
// #include<bits/stdc++.h>
// using namespace std;

// void solve(int arr[], int n){
//        int temp = arr[0];
//        for(int i = 0; i < n - 1; i++){
//           arr[i] = arr[i + 1];
//        }
//        arr[n -1 ] = temp;
//        for(int i = 0; i< n; i++){
//          cout << arr[i] << " ";
//        }
// }

// int main(){
//       int n = 5;
//       int arr[]  = {1, 2, 3, 4,5};
//       solve(arr, n);
// }






//Rotate array by K elements: 
// given an array of integers rotating array of elements by k elements either left or right : 
// #include <bits/stdc++.h>
// using namespace std;

// void Rotatetoright(int arr[], int n , int k ){
//    if(n == 0)
//      return; 
//      k = k % n;
//      if(k > n)
//        return;

//        int temp[k];
//        for(int i = n - k; i<n; i++){
//           temp[i - n + k] = arr[i];
//        }
//        for(int i = n - k - 1; i >=0; i--){
//           arr[i + k] = arr[i];
//        }
//        for(int i = 0; i<k; i++){
//          arr[i] = temp[i];
//        }
// }

// int main(){
//      int n = 7;
//      int arr[] = {1, 2, 3, 4, 5, 6, 7};
//      int k = 2;
//      Rotatetoright(arr, n, k);
//      cout << "After rotating the elements to right " << endl;
//      for(int i = 0; i<n; i++){
//          cout << arr[i] << " ";
//      }
//      return 0;
// }



//for rotating the elements to left: 
// #include<bits/stdc++.h>
// using namespace std;

// void Rotatetoleft(int arr[], int n, int k ){
//  if(n==0)
//    return;
//  k = k % n;
//  if(k > n)
//    return;
//    int temp[k];
//    for(int i = 0; i< k; i++){
//     temp[i] = arr[i];
//    }
//    for(int i = 0; i<n - k; i++){
//      arr[i] = arr[i + k];
//    }
//    for(int i = n - k; i<n; i++){
//       arr[i] = temp[i -n +k];
//    }
// }

// int main(){
//      int n = 7;
//      int arr[] = {1,2,3,4,5,6,7};
//      int k= 2;
//      Rotatetoleft(arr,n,k);
//      cout << "After rotating the elements to left " << endl;
//      for(int i = 0; i< n; i++){
//          cout << arr[i] << " ";
//      }
//      return 0;
// }



// reverse the elements to  right: 
// #include<bits/stdc++.h>
// using namespace std;

// void Reverse(int arr[], int start, int end){
//      while(start <= end){
//           int temp = arr[start];
//           arr[start] = arr[end];
//           arr[end] = temp;
//           start++;
//           end--;
//      }
// }

// void Rotateeletoright(int arr[], int n, int k){
     
//      Reverse(arr, 0, n-k-1);
//      Reverse(arr, n-k, n-1);
//      Reverse(arr, 0, n-1);
// }

// int main(){
//  int arr[] = {1, 2, 3,4,5,6,7};
//  int n = 7;
//  int k = 2;
//  Rotateeletoright(arr, n, k);
//  cout << "After Rotating the k elements to right";
//  for(int i = 0; i<n; i++)
//    cout << arr[i] << " ";
//    cout << endl;
//    return 0;
// }





// rotating elements to left: 
#include<bits/stdc++.h>
using namespace std;

void Reverse(int arr[], int start, int end){
     while(start <= end){
         int temp = arr[start];
         arr[start] = arr[end];
         arr[end] = temp;
         start++;
         end--;
     }
}

void Rotateeletoleft(int arr[], int n, int k){
    Reverse(arr, 0, k - 1);
    Reverse(arr, k , n - 1);
    Reverse(arr, 0, n - 1);
}

int main(){
   int arr[] = {1, 2, 3, 4,5, 6,7};
   int n = 7;
   int k = 2;
   Rotateeletoleft(arr, n, k);
   cout << "After rotating the k elements to left ";
   for(int i = 0; i< n; i++)
     cout << arr[i] << " ";
     cout << endl;
    return 0;

}

File: Array3.js
//Given an array of N integers left rotate the array by one place : brute force approach :

// function solve(arr, n){
//       let temp = new Array(n);
//       for(let i = 1; i < n; i++){
//           temp[i - 1] = arr[i];
//       }
//       temp[n - 1] = arr[0];
//       for(let i = 0; i < n; i++){
//           console.log(temp[i] + " ");
//       }
//       console.log();
// }

// let n = 5;
// let arr = [1,2,3,4,5];
// solve(arr, n);



//optimal approach : 
 
//   function solve(arr, n){
//    let temp = arr[0];
     
//    for(let i = 0; i < n-1; i++){
//       arr[i] = arr[i+1];
//    }
//    arr[n - 1] = temp;
//    for(let i = 0; i < n; i++){
//        console.log(arr[i] + " "); 
//    }
//   }

//   let n = 5;
//   let arr = [1, 2, 3, 4, 5];
//   solve(arr, n);









// 
//Rotate array by K elements: 
// given an array of integers rotating array of elements by k elements either left or right : 
// function rotateToRight(arr, k){
//   const n =arr.length;
//   if(n===0) return;

//   k = k % n;
//   if(k > n) return;

//   const temp = arr.slice(n - k, n);

//   for(let i = n-k-1; i>=0; i--){
//      arr[i + k] =arr[i];
//   }

//   for(let i =0; i< k; i++){
//      arr[i] = temp[i];
//   }
// }

// const arr = [1,2,3,4,5,6,7];
// const k = 2;
// rotateToRight(arr, k);
// console.log("After rotating the elements to right");
// console.log(arr);






//rotating elements to left : 
// function rotateToLeft(arr, k){
//       const n = arr.length;
//       if(n === 0)return;

//       k = k % n;
//       if(k > n) return;

//       const temp = arr.slice(0, k);

//       for(let i = 0; i< n - k; i++){
//           arr[i] = arr[i + k];
//       }
//       for(let i = 0; i<k; i++){
//           arr[n - k + i] = temp[i];
//       }
// }

// let arr1 = [1, 2, 3, 4, 5, 6, 7];
// let k1 = 2;
// rotateToLeft(arr1, k1);
// console.log("After rotating the elements to the left");
// console.log(arr1);




// Rotating elements to the right using reverse method: 
// function reverse(arr, start , end){
//       while(start < end){
//           let temp = arr[start];
//           arr[start] = arr[end];
//           arr[end] = temp;
//           start++;
//           end--;
//       }
// }

// function rotateToRight(arr, k){
//       const n = arr.length;
//       k = k % n;
//       if(n=== 0 || k === 0) return;
//       reverse(arr, 0, n-k-1);
//       reverse(arr, n - k, n-1);
//       reverse(arr, 0, n - 1);
// }

// let arr2 = [1, 2, 3,4,5,6,7];
// let k2 = 2;
// rotateToRight(arr2, k2);
// console.log("After rotating the elements to the right ");
// console.log(arr2);









// rotating elements to the left using reverse method: 
function reverse(arr, start, end){
      while(start < end){
          let temp = arr[start];
          arr[start] =arr[end];
          arr[end] = temp;

          start++;
          end--;
      }
}

function rotateToLeft(arr, k){
      const n = arr.length;
      k =k % n;
      if(n === 0 || k === 0) return;

      reverse(arr,0, k - 1);
      reverse(arr, k , n - 1);
      reverse(arr, 0, n - 1);
}

let arr3 = [1, 2, 3, 4, 5, 6, 7];
let k3 = 2;
rotateToLeft(arr3, k3);
console.log("After rotating the elements to the left");
console.log(arr3);
